### ISSUE TITLE
False positive `no-matching-overload` under certain circumstances

### ISSUE THREAD

**Issue by denialhaag** [astral-sh/ty#3480]
#### MRE

Consider the following `example.pyi`:

```py
from collections.abc import Iterable
from typing import overload

class Target: ...

class ExampleClass:
    @overload
    def append(
        self,
        name: str,
        targets: int | Target | Iterable[int | Target],
        arg: float | Iterable[float],
    ) -> None:
        ...
    @overload
    def append(
        self,
        name: str,
    ) -> None:
        ...
    def append(
        self,
        name: object,
        targets: object = (),
        arg: object = None,
    ) -> None:
        ...
```

Consider also the following user code:

```py
from example import ExampleClass

def main() -> None:
    example = ExampleClass()
    example.append("test", [1, 2, 3])
```

Running `ty check` raises the following error:

```
error[no-matching-overload]: No overload of bound method `ExampleClass.append` matches arguments
  --> example.py:13:5
   |
13 |     example.append("test", [1, 2, 3])
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
info: First overload defined here
  --> example.pyi:14:5
   |
14 | /     @overload
15 | |     def append(
16 | |         self,
17 | |         name: str,
18 | |         targets: int | Target | Iterable[int | Target],
19 | |         arg: float | Iterable[float],
20 | |     ) -> None:
21 | |         ...
   | |___________^ First overload defined here
   |
info: Possible overloads for bound method `append`:
info:   (self, name: str, targets: int | Target | Iterable[int | Target], arg: int | float | Iterable[int | float]) -> None
info:   (self, name: str) -> None
info: Overload implementation defined here
  --> example.pyi:28:9
   |
28 |       def append(
   |  _________^
29 | |         self,
30 | |         name: object,
31 | |         targets: object = (),
32 | |         arg: object = None,
33 | |     ) -> None:
   | |_____________^
   |

Found 1 diagnostic
```

The `ty` configuration:

```toml
[tool.ty.terminal]
error-on-warning = true

# This obviously doesn't change anything, but I'll include it anyway:
[tool.ty.src]
exclude = [
    "**/*.ipynb",
    "docs/**",
    "scripts/**",
]
```

#### Observation

Interestingly, I do not get this error if I remove `arg: float | Iterable[float]` and `arg: object = None` from `examples.pyi`, even though the `arg` argument is not involved in the example.

### Version

ty 0.0.36 (a63e55929 2026-05-14)

**Comment by charliermarsh** [comment]
This looks correct to me because the first overload signature doesn't indicate that `arg` is optional:
```python
@overload
def append(
    self,
    name: str,
    targets: int | Target | Iterable[int | Target],
    arg: float | Iterable[float],
) -> None:
    ...
```

If it's optional, you should encode that in the overload definition (the first `def`), not just the implementation (the last). Did you mean to do something like `arg: None | float | Iterable[float] = None`, or `example.append("test", [1, 2, 3], 1.0)` at the usage site?

**Comment by denialhaag** [comment]
Ohh wow, I've looked at this for the better part of an hour, but I must have missed that detail... `arg` is really meant to be optional. Thanks a lot for pointing this out, @charliermarsh! 🙂

I'll report this to the library distributing the stubs.

Closing this.
